### DISCUSSION TITLE
Streamlit and role grants

### DISCUSSION THREAD

**Discussion by kokorin** [littleK0i/SnowDDL#162] [Ideas]
In one of our PoC we wanted to use SnowDDL to create a DB and schema to create Streamlit app. 
Streamlit app itself will be deployed outside of SnowDDL so we don't need to have `StreamlitBlueprint` in this project.

First, I have to describe the problem: Snowflake doesn't support future grants on Streamlit, but it can grant privileges on existing ones.

We want to use SnowDDL to grant USAGE privilege on Streamlit to reader role. For that we even added STREAMLIT object type into SnowDDL (by modifying py-files in venv) like this:

```py
class ObjectType(Enum):
    # ...
    STREAMLIT = {
        "singular": "STREAMLIT",
        "plural": "STREAMLITS",
    }
    # ...
```

And then added Streamlit to Permission Model:

```yml
out_model:
  inherit_from: default
  ruleset: SCHEMA_OWNER
  owner_create_grants:
    - STREAMLIT
  read_future_grants:
    STREAMLIT: [USAGE] 
```

Unfortunately, `snowddl plan` suggests a query to remove grants on existing Streamlit app from a Reader Role (we granted it manually).

Then we added `is_future_grant_supported` to Streamlit:
```py
class ObjectType(Enum):
    # ...
    STREAMLIT = {
        "singular": "STREAMLIT",
        "plural": "STREAMLITS",
        "is_future_grant_supported": True,
    }
    # ...
```
After that SnowDDL doesn't suggest to revoke permissions on Streamlit, but it instead suggests to grant permissions on future Streamlits, which is not possible in Snowflake ATM.

As a workaround for the PoC we decided to create a dedicated DB and use DB Owner Role to create/update Streamlit. A Business Role (which is granted with Owner Role) is granted with Read Role on another DB where actual data is stored, so for now it's enough for use to make sure that a BR can't occasionally modify data but can modify Streamlit.

Now the question: is it possible to Streamlit object type to SnowDDL with support of grants `ON ALL SRTEAMLITS` but without `ON FUTURE STREAMLITS`.

My idea is to add another property to ObjectType to control such behavior:
```py
class ObjectType(Enum):
    # ...
    STREAMLIT = {
        "singular": "STREAMLIT",
        "plural": "STREAMLITS",
        "is_all_grant_supported": True,
        "is_future_grant_supported": False,
    }
    # ...
    @property
    def is_all_grant_supported(self) -> bool:
        return self.value.get("is_all_grant_supported", self.is_future_grant_supported) 
```

**Comment by littleK0i** [comment]
In theory, global roles can be used to cover almost any edge case.

Since Streamlit app is managed outside of SnowDDL, it might be ok to also maintain a global role with grants for all streamlits. This global role can be granted to schema owner role or business role.

Btw, I also heard a lot of other negatives about Snowflake native Streamlit implementation. Lack of future grants is not the only problem. As far as I know, it is also outdated in terms of version and overpriced, because it requires a warehouse to run.

Self-hosting might be easier than trying to find work-arounds for Snowflake laziness. Also, there are lots of better products on the market, e.g. [Hex](https://hex.tech/).

**Comment by littleK0i** [comment]
Consider an option with Streamlit running in container: https://www.reddit.com/r/snowflake/comments/1d5kwl5/why_streamlit_in_native_apps_is_so_old_and_has_so/

Maybe it can help to alleviate most of known problems, including a problem with unfinished implementation of permissions.

**Comment by DoMy91** [comment]
It seems that Snowflake now supports future grants on Streamlits: https://docs.snowflake.com/en/release-notes/streamlit-in-snowflake#january-31-2025-support-for-future-grants-in-sis-general-availability

**Reply by littleK0i** [reply]
Does not seem to work for ownership, at least not on my test account.

```
(
    message   =>  Unsupported feature GRANT/REVOKE OWNERSHIP ON STREAMLIT'.
    errno     =>  391811
    sqlstate  =>  0A000
    sfqid     =>  01ba35ca-010c-25eb-0002-30d301b997fa
    sql       =>  GRANT OWNERSHIP ON FUTURE STREAMLITS IN SCHEMA "PYTEST__STREAMLIT_DB1"."STREAMLIT_SC1" TO ROLE "PYTEST__STREAMLIT_DB1__STREAMLIT_SC1__OWNER__S_ROLE"
)
```

**Reply by kokorin** [reply]
@littleK0i do you think it's possible to add support of `STREAMLIT` object type in Technical Roles? We use workaround with global roles at the moment, but it becomes problematic to manage global roles.

**Reply by littleK0i** [reply]
@kokorin , in order for object type to work with technical roles we also need a "blueprint" class. It is used to understand if we need to add "env_prefix" or not. Also, technical roles can only have reliable grants on objects created by SnowDDL itself.

So "global role" is probably the right solution for now. At least until Snowflake implements missing ownership grants.

**Comment by y-tee** [comment]
hello, any plans to implement streamlit in snowddl now?

**Reply by littleK0i** [reply]
As far as I know, Snowflake effectively abandoned Streamlit, and none of previously known issues were resolved.

Issues with Snowflake-native Streamlit are:
- Outdated version, many limitations.
- Must use warehouse to run, which is expensive.
- Future grants are not supported.

The current advice is:
1) Use something better and cheaper, like [Hex](https://hex.tech/).
2) If you absolutely must use Streamlit, run it outside of Snowflake. It provides full freedom, eliminates all limitations, reduces cost to peanuts.

Currently no plans to implement it for SnowDDL, since there is no reason to use this feature.
